corehq/apps/users/signals.py

```diff
@@ -30,13 +31,20 @@ def django_user_post_save_signal(sender, instance, created, raw=False, **kwargs)
     return CouchUser.django_user_post_save_signal(sender, instance, created)
 
 
+def _should_sync_to_es():
+    # this method is useful to disable update_user_in_es in all tests
+    #   but still enable it when necessary via mock
+    return not settings.UNIT_TESTING
```

Comment:
It is best practice to keep test code out of production code. To achieve that here, it we could do something like

Put this somewhere like `corehq/util/test_utils.py`
```py
from contextlib import contextmanager
from mock import patch

_sync_user_to_es_patch = patch('corehq.apps.users.signals.send_to_elasticsearch')
_sync_user_to_es_patch.start()

@contextmanager
def sync_users_to_es():
    """Contextmanager/decorator to enable sync users to ES

    Usage as decorator:

        @sync_users_to_es()
        def test_something():
            ...  # do thing with users in ES

    Usage as context manager:

        with sync_users_to_es():
            ...  # do thing with users in ES
    """
    _sync_user_to_es_patch.stop()
    try:
        yield
    finally:
        _sync_user_to_es_patch.start()
```

Reply:
Thanks for the very clear comment, Daniel! The only concern is this adds a slight bit of magic where if someone were to debug a test where they expect the user to be synced, it would be a bit difficult to find this out by just looking at `update_user_in_es`, because this patches that method. This is an okay tradeoff, you think? 

Additionally the non-default behaviour (of requiring user to be synced to ES) is so rarely required in tests that a decorator like this is not super necessary?

Reply:
> This is an okay tradeoff, you think?

Yes, I think it's more important to keep test code out of production code.

> Additionally the non-default behaviour (of requiring user to be synced to ES) is so rarely required in tests that a decorator like this is not super necessary?

The decorator is simply a replacement for
```py
@patch('corehq.apps.users.signals._should_sync_to_es', return_value=True)
```

Reply:
Beautiful yet magical. Do we flag such changes such that devs are aware of it?
If as practice we enforce
> keep test code out of production code

It will become more of a habit and hence people would be more aware of it.
If i had not seen this comment, I would have definitely ended time trying to debug, if i was testing an elastic search change.

Reply:
> Yes, I think it's more important to keep test code out of production code.

I understand the usefullness of keeping those things separate, but I feel it's more important to avoid code that does magic. But yeah, I do not have enough experience to gauge which one of this is a better tradeoff. At this instance, so far I don't have any ideas to do this without any magic. Any other ideas to do this without the magic? @millerdev  or does others also think this is an okay tradeoff @snopoke @kaapstorm ?

Reply:
In the interest of fixing the build and since this PR has been approved I am going to merge this for now, and address above comment in a separate PR

Reply:
I agree with @millerdev that we should make a good effort to keep test code out of production code. The pattern of using `settings.UNIT_TESTING` is one that is used in a number of other places and is similar to the pattern of using `settings.DEBUG` which is also used in a number of places.

I do like @millerdev solution however I'm not quite sure how we would ensure that the patch is only ever applied when in `test mode`. I guess it could be made a part of the test runner but that does also hide it away and make it hard to find.

Overall I think it's OK to use this pattern but any alteration to code that makes it different in tests can cause confusion. This code in particular is quite complex:

* only fire signals if `fire_signals == True`
* write to ES in a signal handler but only if `_should_sync_to_es()`

So now there are 2 conditions which must be met in order to update the user in ES.

I'm not suggesting any changes but just making a note.

Reply:
I had forgotten that `coreqh.util.test_utils` is imported by production code (we should fix that), and therefore would cause the patch to be applied in production. **That's really bad!!** In that case I would suggest putting the patch and decorator code in a new module.

> Any other ideas to do this without the magic @millerdev?

Here's a PR: https://github.com/dimagi/commcare-hq/pull/27894

In the process of moving it to a new module I also devised a more reliable way to have the patch applied only when running tests (using `coreqh.tests.noseplugins.patches`) and also added assertions to verify that the patches are only applied when running tests.